zero/src/prover.rs

```diff
@@ -131,29 +135,180 @@ impl BlockProverInput {
+                    // Mark the end of the batch segments by sending `None`
+                    segment_tx
+                        .send(None)
+                        .await
+                        .context(format!("failed to send end segment data indicator for batch {batch_idx}"))?;
+                    Ok::<(), anyhow::Error>(())
+                });
+            }
+            while let Some(it) = batch_segment_futures.next().await {
+                // In case of an error, propagate the error to the main task
+                it?;
+            }
+            Ok::<(), anyhow::Error>(())
+        });
+
+        let proof_runtime_ = proof_runtime.clone();
+        let batches_proving_task = tokio::spawn(async move {
+            let mut batch_proving_futures: FuturesUnordered<_> = FuturesUnordered::new();
+            // Span a proving subtask for each batch where we generate segment proofs
+            // and aggregate them to batch proof.
+            for (batch_idx, mut segment_rx) in segment_receivers.into_iter().enumerate() {
+                let batch_proof_tx = batch_proof_tx.clone();
+                let seg_prove_ops = seg_prove_ops.clone();
+                let seg_agg_ops = seg_agg_ops.clone();
+                let proof_runtime = proof_runtime_.clone();
+                // Tasks to dispatch proving jobs and aggregate segment proofs of one batch
+                batch_proving_futures.push(async move {
+                    let mut batch_segment_aggregatable_proofs = Vec::new();
+
+                    // This channel collects segment proofs from the one batch
+                    // proven in parallel. The size of this channel does not matter much,
+                    // as it is only used to collect segment aggregatable proofs.
+                    let (segment_proof_tx, mut segment_proof_rx) =
+                        mpsc::channel::<(usize, crate::proof_types::SegmentAggregatableProof)>(32);
+
+                    // Wait for segments and dispatch them to the segment proof worker task.
+                    // The segment proof worker task will prove the segment and send it back.
+                    let mut segment_counter = 0;
+                    let mut segment_proving_tasks = Vec::new();
+                    while let Some(segment_data) = segment_rx.recv().await {
+                        if segment_data.is_none() {
+                            break;
+                        }
+                        let seg_prove_ops = seg_prove_ops.clone();
+                        let proof_runtime = proof_runtime.clone();
+                        let segment_proof_tx = segment_proof_tx.clone();
+                        // Prove one segment in a dedicated async task.
+                        let segment_proving_task = tokio::spawn(async move {
+                            if let Some(segment_data) = segment_data {
+                                debug!("proving the batch {batch_idx} segment nr. {segment_counter}");
```

Comment:
nit: [logging](https://0xpolygon.slack.com/archives/C073HA1NSDU/p1727888673088639)
```suggestion
                                debug!(%batch_idx, %segment_counter, "proving batch segment");
```